feat: store ABI file and commit hash of each deploy

0xble opened the pull request:
### Motivation

Help FE track ABI changes.

### Solution

-  Store the latest commit for each deployment.
-  Store the ABI file for each deployment under `deployments/abi/[CONTRACT]/v[VERSION].json`.
- Notify team whenever ABI changes between contract versions.
- Include link to Gist of diff between previous and latest version ABIs in notification.
- Requires adding GitHub token to repo as a secret to upload the Gist.
- Requires adding Discord webhook URL as a secret to send the notification.

I considered the following features but decided against them:
-  Storing .diff files for each contract version update to compare changes between the latest and previous ABIs. This felt too opinionated. For example, what if someone wanted to view changes between the latest and the third-to-last commit? Instead, I opted to store each ABI file, searchable by contract and version.

0xble committed:
store ABI file and commit hash of each deploy

arr00 commented on src/deploy.ts:
```diff
@@ -153,24 +151,23 @@ async function runDeploy(
     .digest("hex");
+  newDeploy.commitHash = getLatestCommitHash();
+
+  // Store ABI file
+  const abiDir = `deployments/abi/${contract}`;
+  if (!fs.existsSync(abiDir)) {
+    fs.mkdirSync(abiDir, { recursive: true });
+  }
+  fs.writeFileSync(
+    `${abiDir}/v${newDeploy.version.replace(/\./g, "_")}.json`,
+    JSON.stringify(contractJson.abi, null, 2),
+  );
```
This should be option. Lets add another flag like `storeAbi` for the CLI.

arr00 commented on src/deploy.ts:
```diff
@@ -519,3 +517,16 @@ function getProjectContracts(): string[] {
+ */
+function getLatestCommitHash(): string {
+  try {
+    return execSync("git rev-parse HEAD").toString().trim();
```
Should work in CI but we should test before pushing a release

0xble replied:
Will definitely test before merging

arr00 reviewed (changes requested):
Looks good--nice move adding the `contractJson` const. We should test in CI.

0xble committed:
notify in Discord of ABI changes

0xble committed:
make writing ABIs optional

0xble committed:
display network name instead of chain ID

0xble committed:
style

arr00 commented:
What is the empty `constants.ts` file?

arr00 commented on src/deploy.ts:
```diff
@@ -153,24 +166,34 @@ async function runDeploy(
+    if (!fs.existsSync(abiDir)) {
+      fs.mkdirSync(abiDir, { recursive: true });
+    }
+    const newAbiPath = `${abiDir}/v${newDeploy.version.replace(/\./g, "_")}.json`;
```
How does this fit into minor version changes?

0xble replied:
Ah good catch

arr00 reviewed (commented):
Some thoughts after a more detailed review.

arr00 commented on src/deploy.ts:
```diff
@@ -519,3 +543,170 @@ function getProjectContracts(): string[] {
+  }
+
+  const newAbiPath = `${abiDir}/${newVersion.replace(/\./g, "_")}.json`;
+  const previousAbiPath = `${abiDir}/${files[files.length - 2]}`;
```
https://github.com/PartyDAO/solidity-deploy/blob/a9c24693c4045b6ba9886347df1774c6dd560ec0/src/deploy.ts#L578

0xble committed:
Delete src/constants.ts

0xble committed:
sort abi files

0xble committed:
remove Discord-specific wording

0xble committed:
refactor handling if no ABI changes

0xble commented:
Going to test this, will request review after confirming it works as expected

0xble committed:
Merge branch 'main' of https://github.com/PartyDAO/solidity-deploy into feat/store-abi-and-commit

0xble commented:
Confirmed it works
![zdLGlajg](https://github.com/user-attachments/assets/e4a49c32-4b5b-4602-8ec8-6d4f06c2a2ec)

0xble committed:
Merge branch 'main' of https://github.com/PartyDAO/solidity-deploy into feat/store-abi-and-commit

0xble committed:
fix bugs

0xble commented:
Tested in a separate repo, confirmation it works
![zdLGlajg](https://github.com/user-attachments/assets/e4a49c32-4b5b-4602-8ec8-6d4f06c2a2ec)

arr00 commented on package.json:
```diff
@@ -7,12 +7,17 @@
     "@types/node": "^20.14.5",
+    "@types/semver": "^7.5.8",
     "@types/yargs": "^17.0.32",
+    "axios": "^1.7.4",
+    "diff": "^5.2.0",
```
Aren't these not dev dependencies?

arr00 commented:
Seems like this PR reverts changes from #6.

0xble commented:
> Seems like this PR reverts changes from #6.

I'll fix that

arr00 commented on src/deploy.ts:
```diff
@@ -194,9 +197,73 @@ async function runDeploy(
+      )
+
+      // Check if there's a previous version
+      if (versions.length > 0) {
```
shouldn't this be > 1?

0xble committed:
suggestions

0xble committed:
suggestions

- Add notification for new deploys with unchanged ABI
- Restore accidently deleted error handling in `validateDeploy()`

arr00 reviewed (commented):
looks good. some final nits

0xble committed:
style: rename `contractAddress` to `address` in `notifyNewDeploy()`

0xble committed:
chore: replace `axios.post()` with `sendWebhookMessage()` function

0xble commented on src/deploy.ts:
```diff
@@ -534,3 +613,242 @@ function getProjectContracts(): string[] {
+  filename: string,
+  description: string,
+  githubToken: string
+): Promise<string> {
```
Not sure what's wrong here. Async functions always return Promises

arr00 replied:
ah yes my bad never-mind.
